text/0000-snapshot-system.md

```diff
@@ -0,0 +1,207 @@
+
+Dependencies between variable updates and spawning will be managed by the Snapshot System. You will only ever get updates to variables belonging to objects that were spawned. State information (NetworkVariable) is updated with latest available value, and intermediate values can be skipped. That happens, for example, if there's packet re-ordering in transit.
+
+Event information (spawns) is guaranteed to be applied only once, and in order. 
+
+(1) When we refer to retransmit, we mean that further Snapshot Messages will carry the same information again. Past Snapshot Messages will not be re-sent identically.
+
+### Enabling Snapshot
+
+Whether Snapshot is enabled or not, for NetworkVariables or Spawns, should have no visible impact to the user, apart for:
+- bandwidth improvements
+- delivery improvements
+- consistency between delivery of NetworkVariable updates and Spawning
+
+To enable or disable it, the following public variables will be offered in `NetworkConfig`:
+
+```
+public bool UseSnapshotVariable = false;
+bublic bool UseClassicSpawn = false;
+```
+
+The first one will control whether NetworkVariables are sent via the Snapshot System or not, and the second one whether the Spawns are sent via the Snapshot System.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+### Guiding principles
+
+The first guiding principle to introduce is that under normal operations **no** memory allocations are done at runtime, within the Snapshot System. Every buffer is pre-allocated when the Snapshot System is started. Provisioning is made for sending, for receiving, and for any internal processing.  
+
+Another guiding principle is that the Snapshot System doesn't care whether we have a star topology, a full mesh, or anything in between. Rather it is only concerned about:
+- The information it has about Variables and Spawns at various points in time
+- Which information was acknowledged by each connection
+- What information is received on a connection
+
+We will still distinguish, in the description below, between:
+- a machine that is connected to a single other. We'll call that a client
+- a machine that is connected to many other machine. We'll call that a host
+
+A third principle is to never try to capture the whole world state directly. It should be possible to query every variable at any tick, but it should not be necessary to deal with a message that contains all variables. Thus, at scale, NetworkVariables that are not touched are left out of the snapshots.
+
+### Maintained data
+
+In the example above, we'll assume two clients, Alice and Bob, connected to a host, Charlie.
+
+In order to perform its job, the Snapshot System will maintain the following data: (the usage of this data will be presented later)
+
+On hosts, only, for each connection (client):
+- List of interesting objects (AoI-provided)
+
+On all machines, for each connection:
+- For each object: acknowledged spawned status, with tick.
```

Comment:
Does that include just spawned objects on clients? Or will a client have track over all existing objects in total. Like In a battle royale will the client keep a list of every chest in the entire world?